api/search.go

```diff
@@ -1,28 +1,618 @@
+				}
+
+				if len(res.Hits) >= int(limit) {
+					break
+				}
+			}
+			response.Results[i] = res
+		case "COURSES":
+			hits := res.Hits
+			res.Hits = []any{}
+			response.Results[i] = meilisearch.SearchResponse{}
+
+			var meiliCourses []SearchCourseDTO
+			temp, err := json.Marshal(hits)
+			if err != nil { // shouldn't happen
+				continue
+			}
+			err = json.Unmarshal(temp, &meiliCourses)
+			if err != nil { // shouldn't happen
+				continue
+			}
+
+			for _, meiliCourse := range meiliCourses {
+				course, err := daoWrapper.CoursesDao.GetCourseBySlugYearAndTerm(c, meiliCourse.Slug, meiliCourse.TeachingTerm, meiliCourse.Year)
+				if err == nil && user.IsEligibleToSearchForCourse(course) {
+					res.Hits = append(res.Hits, meiliCourse)
+				}
+
+				if len(res.Hits) >= int(limit) {
+					break
+				}
+			}
+			response.Results[i] = res
+		case "SUBTITLES":
+			hits := res.Hits
+			res.Hits = []any{}
+			response.Results[i] = meilisearch.SearchResponse{}
+
+			var meiliSubtitles []SearchSubtitlesDTO
+			temp, err := json.Marshal(hits)
+			if err != nil { // shouldn't happen
+				continue
+			}
+			err = json.Unmarshal(temp, &meiliSubtitles)
+			if err != nil { // shouldn't happen
+				continue
+			}
+
+			for _, meiliSubtitle := range meiliSubtitles {
+				stream, err := daoWrapper.StreamsDao.GetStreamByID(c, strconv.FormatUint(uint64(meiliSubtitle.StreamID), 10))
+				if err != nil {
+					continue
+				}
+				course, err := daoWrapper.CoursesDao.GetCourseById(c, stream.CourseID)
+				if err != nil {
+					continue
+				}
+
+				meiliSubtitle.StreamName = stream.Name
+				meiliSubtitle.StreamStartTime = stream.Start
+				meiliSubtitle.StreamEndTime = stream.End
+				meiliSubtitle.CourseSlug = course.Slug
+				meiliSubtitle.CourseName = course.Name
+				meiliSubtitle.CourseYear = course.Year
+				meiliSubtitle.CourseTeachingTerm = course.TeachingTerm
+				if userEligibleToSeeResultsOfHiddenCourse(course) && (!stream.Private || user.IsAdminOfCourse(course)) {
+					res.Hits = append(res.Hits, meiliSubtitle)
+				}
+
+				if len(res.Hits) >= int(limit) {
+					break
+				}
+			}
+			response.Results[i] = res
+		default:
+			continue
+		}
+	}
+}
+
+// meilisearch filter
+
+func meiliSubtitleFilter(user *model.User, courses []model.Course) string {
+	if len(courses) == 0 {
+		return ""
+	}
+
+	var streamIDs []uint
+	for _, course := range courses {
+		admin := user.IsAdminOfCourse(course)
+		for _, stream := range course.Streams {
+			if !stream.Private || admin {
+				streamIDs = append(streamIDs, stream.ID)
+			}
+		}
+	}
+	return fmt.Sprintf("streamID IN %s", uintSliceToString(streamIDs))
+}
+
+func meiliStreamFilter(c *gin.Context, user *model.User, semester model.Semester, courses []model.Course) string {
+	if courses != nil {
+		return fmt.Sprintf("courseID IN %s", courseSliceToString(courses))
+	}
+
+	semesterFilter := fmt.Sprintf("(year = %d AND semester = \"%s\")", semester.Year, semester.TeachingTerm)
```

Comment:
This looks awfully a lot like SQL injection would be possible here? Please correct me if I'm wrong, but ugh.

Reply:
I'm currently on vacation and will address this in detail either tomorrow or next week. Short answer: User/client input is checked, so Injections are not possible, if these checks work as intended.

Reply:
This function expects all parameters to be "friendly". This means that the slice `courses` may only consist of courses that the user is allowed to search through and all semesters consisting of a base-10 four digit `year` and the `teachingTerm` being either "W" or "S" and nothing else. I added comments to clarify that these checks are the caller's responsibility.

When looking at the usages, the slice `courses` is either `nil` or the return value of `parseCourses`, which retrieves the returned courses from the database. If there's an SQL injection possible up to this point, there'd be another problem. The returned courses are then checked here:
https://github.com/TUM-Dev/gocast/blob/94dae0434a527817865a9942acd120b2cece3348/api/search.go#L100-L105

Similarly the `semester` is either of zero value or the return value of `parseSemesters`. `parseSemesters` ensures that the semesters from the URL parameter conform to the format mentioned at the beginning using this regex
https://github.com/TUM-Dev/gocast/blob/94dae0434a527817865a9942acd120b2cece3348/api/search.go#L481

Am I missing something or misunderstanding you?

Reply:
However, there's another mistake here:
https://github.com/TUM-Dev/gocast/blob/94dae0434a527817865a9942acd120b2cece3348/api/search.go#L352-L354 This ignores the case, where the user is not allowed to see private streams. However, this mistake does not lead to data leaks, which is why it did not come up during testing. Even if Meilisearch returns streams the user is not allowed to see, this results will get thrown out by `checkAndFillResponse`.
I'll fix it though.

Reply:
Regarding the latest comment: See https://github.com/TUM-Dev/gocast/pull/1380#discussion_r1781055089

Reply:
> if these checks work as intended.

Yea... I really don't like seeming query construction using string formatting regardless of prior checks, but it seems like (please correct me if I'm wrong) mailisearch-go does not have prepared statements, which is utterly rediculous.

Reply:
Makes sense.
> but it seems like (please correct me if I'm wrong) mailisearch-go does not have prepared statements
To my knowledge that is correct.